numpy/core/src/multiarray/shape.c

```diff
@@ -246,7 +247,7 @@ PyArray_Newshape(PyArrayObject *self, PyArray_Dims *newdims,
             if (newcopy == NULL) {
                 return NULL;
             }
-            incref = NPY_FALSE;
+            Py_DECREF(self);
```

Comment:
What happens when `newcopy == NULL`? The Py_DECREF should be before line 247

Reply:
There are other places where the code exits early with `return NULL`, so probably better to `goto fail` here and in the other ones as well

Reply:
Same thing as what happened before this patch. Good catch though - I guess this now needs a bug label too.